source/adapters/level_zero/queue.hpp

```diff
@@ -30,6 +31,124 @@ extern "C" {
 ur_result_t urQueueReleaseInternal(ur_queue_handle_t Queue);
 } // extern "C"
 
+struct ur_completion_batch;
+using ur_completion_batch_list = std::list<ur_completion_batch>;
+using ur_completion_batch_it = ur_completion_batch_list::iterator;
+
+// Event completion batch for aggregating status checks of many events into
+// a single one through a barrier. Batches can be continuously reused.
+// Batches start empty and accumulate events, get sealed (which issues
+// an asynchronous barrier on the command list), and then must be waited on
+// for all the events to complete.
+struct ur_completion_batch {
+  ur_completion_batch();
+  ~ur_completion_batch();
+
+  enum state {
```

Comment:
nit: I think it would be simpler to just have separate containers for empty, accumulating, sealed, and completed batches. Then, a state transition would be just moving a batch from one container to another. All functions that expect a batch to be in a certain state would just have to iterate over a single container.

Reply:
I considered doing something similar, with a proper algebraic data type where the implementation for the batch differs depending on its state. Ultimately, I think in C++ it's too much work for not much benefit...